docs/design/metainfo-rpc.md

```diff
@@ -26,87 +26,211 @@ The following is a proto file for just the rpc portion to create segments. The r
+    Status status;
+
+    bytes  stream_id;
+
+    google.protobuf.Timestamp created_at;
+    google.protobuf.Timestamp status_at;
+    google.protobuf.Timestamp expires_at;
+
+    bytes  encrypted_metadata_nonce;
+    bytes  encrypted_metadata;
+
+    int64                fixed_segment_size;
+    RedundancyScheme     redundancy_scheme;
+    EncryptionParameters encryption_parameters;
+
+    int64 total_size;
+    int64 inline_size;
+    int64 remote_size;
 }
 
-message StartStreamAndStartSegmentResponse {
-    // Start Stream
-    bytes stream_upload_id = 1;
+message ObjectCreateRequest {
+    bytes  bucket = 1;
+    bytes  encrypted_path = 2;
+    int32  version = 3;
+}
+
+message ObjectCreateResponse {
+    bytes  bucket;
+    bytes  encrypted_path;
+    int32  version;
+
+    bytes  stream_id;
+
+    google.protobuf.Timestamp expires_at;
+
+    bytes  encrypted_metadata_nonce;
+    bytes  encrypted_metadata;
+
+    RedundancyScheme     redundancy_scheme;
+    EncryptionParameters encryption_parameters;
+}
 
-    // Start Segment
-    bytes segment_upload_id = 2;
-    repeated AddressedOrderLimit addressed_limits = 3;
-    bytes root_piece_id = 4 [(gogoproto.customtype) = "PieceID", (gogoproto.nullable) = false];
+message ObjectCommitRequest {
+    bytes  bucket = 1;
+    bytes  encrypted_path = 2;
+    int32  version = 3;
 }
 
-message CommitSegmentAndStartSegmentResponse {
-    // Commit Segment
-    pointerdb.Pointer pointer = 1;
+message ObjectCommitResponse {
+    Object object = 1;
+}
+
+message ObjectListRequest {
+    bytes     bucket = 1;
+    bytes     encrypted_prefix = 2;
+    bytes     encrypted_cursor = 3;
+    Direction direction = 4;
+    int32     limit = 5;
+    bool      recursive = 6;
+
+    fixed32   meta_flags = 7;
+    bool      include_partial = 8;
+    bool      include_all_versions = 9;
+}
+
+message ObjectListResponse {
+    repeated Object items = 1;
+    bool more = 2;
+}
+
+message ObjectDeleteRequest {
+    bytes  bucket;
+    bytes  encrypted_path;
+    int32  version;
+}
+
+message ObjectDeleteResponse {
+    ???
+}
+
+message ObjectFinishDeleteRequest {
+    bytes  bucket;
+    bytes  encrypted_path;
+    int32  version;
+}
+
+message ObjectFinishDeleteResponse {}
+
 
-    // Start Segment
-    bytes segment_upload_id = 2;
-    repeated AddressedOrderLimit addressed_limits = 3;
-    bytes root_piece_id = 4 [(gogoproto.customtype) = "PieceID", (gogoproto.nullable) = false];
+message SegmentCreateRequest {
+    bytes stream_id;
+    int32 part_number;
+    int32 index;
+
+    int64 max_encrypted_segment_size;
+}
+
+message SegmentCreateResponse {
+    repeated AddressedOrderLimit addressed_limits;
 }
 
-message CommitSegmentAndCommitStreamResponse {
-    // Commit Segment
-    pointerdb.Pointer pointer = 1;
+message AddressedOrderLimit {
+    orders.OrderLimit2 limit   = 1;
+    node.NodeAddress   address = 2;
+}
+
+message SegmentCommitRequest {
+    bytes stream_id;
+    int32 part_number;
+    int32 index;
```

Comment:
Similarly, I would prefer if this used a segment_upload_id or something to allow the server to use it like a browser cookie.

Reply:
But that commits to the cookie's format. If it's bytes, we have room to insert information into that cookie that clients that we may not be able to update don't have to understand to return to us.

Reply:
Yeah, but I cannot see what that exact information would be. If there's more information here then a lot of the design would probably change as well... 

But, having better compatibility isn't worse and implementation wouldn't be much more difficult. So `segment_id` it is.